File: android/src/main/java/com/witsystem/top/flutterwitsystem/add/ble/AddBleDevice.java
package com.witsystem.top.flutterwitsystem.add.ble;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCallback;
import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattDescriptor;
import android.bluetooth.BluetoothGattService;
import android.bluetooth.BluetoothProfile;
import android.content.Context;
import android.util.Log;

import com.witsystem.top.flutterwitsystem.add.AddDevice;
import com.witsystem.top.flutterwitsystem.ble.Ble;
import com.witsystem.top.flutterwitsystem.ble.BleCode;
import com.witsystem.top.flutterwitsystem.device.DeviceManager;
import com.witsystem.top.flutterwitsystem.net.HttpsClient;
import com.witsystem.top.flutterwitsystem.tools.ByteToString;
import com.witsystem.top.flutterwitsystem.tools.NetWork;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Timer;
import java.util.TimerTask;
import java.util.UUID;

/**
 * 添加蓝牙设备
 */
public class AddBleDevice extends BluetoothGattCallback implements AddDevice, BluetoothAdapter.LeScanCallback {

    private AddBleDeviceCall addBleDeviceCall;

    private Context context;

    private String appId;

    private String token;

    private static AddDevice addDevice;

    private Timer timer;

    //安全认证通过
    private static final int SECURITY_OK = 0x00;

    //安全认证失败
    private static final int SECURITY_FAIL = 0x01;

    //沒有进入设置状态
    private static final int SECURITY_NO_SETTING_STATE = 0x02;

    //网络认证的权限码
    private int checkCode = 0;

    //读取出来的设备信息
    private DeviceInfo deviceInfo;


    private AddBleDevice(Context context, String appId, String token) {
        this.context = context;
        this.appId = appId;
        this.token = token;
    }

    public static AddDevice instance(Context context, String appId, String token) {
        if (addDevice == null) {
            addDevice = new AddBleDevice(context, appId, token);
        }
        return addDevice;
    }

    @Override
    public void addCall(AddBleDeviceCall addBleDeviceCall) {
        this.addBleDeviceCall = addBleDeviceCall;
    }


    @Override
    public void scanDevice() {
        scan();
    }

    @Override
    public void stopDevice() {
        Ble.instance(context).getBlueAdapter().stopLeScan(this);
        processCall(null, BleCode.SCAN_END);
    }

    @Override
    public void addDevice(String deviceId) {
        //???ID的转成mac
        if (deviceId == null) {
            return;
        }
        BluetoothDevice remoteDevice = Ble.instance(context).getBlueAdapter().getRemoteDevice(deviceId);
        connection(remoteDevice);
    }


    private void scan() {
        BluetoothAdapter blueAdapter = Ble.instance(context).getBlueAdapter();
        if (!blueAdapter.isEnabled()) {
            errorCall(null, "Bluetooth not on", BleCode.DEVICE_BLUE_OFF);
            return;
        }
        boolean startLeScan = blueAdapter.startLeScan(this);
        if (!startLeScan) {
            errorCall(null, "Bluetooth not on", BleCode.BLE_SCAN_FAIL);
            stopDevice();
            return;
        }
        processCall(null, BleCode.SCANNING);
    }

    //------------------蓝牙扫描的回调--------------------------------//
    @Override
    public void onLeScan(BluetoothDevice device, int rssi, byte[] scanRecord) {
        Log.d("扫描到的设备", device.getName() + "::" + ByteToString.bytesToHexString(scanRecord));
        if ("Slock".contains(device.getName()) && scanRecord[5] == -15 && scanRecord[6] == -1) {
            processCall(device.getAddress(), BleCode.SCAN_ADD_DEVICE_INFO);
            scanDeviceCall(device.getName(), rssi);
        }
    }


    /**
     * 连接设备
     */
    private void connection(BluetoothDevice device) {
        stopDevice();
        processCall(device.getName(), BleCode.CONNECTING);
        List<BluetoothDevice> connectedDevices = Ble.instance(context).getBluetoothManager().getConnectedDevices(BluetoothProfile.GATT_SERVER);
        if (connectedDevices.toString().contains(device.getAddress())) {
            errorCall(device.getAddress(), "Another app of the phone is connected to the device", BleCode.OTHER_APP_CONN_DEVICE);
        } else {
            BluetoothGatt gatt = device.connectGatt(context, false, this);
            startTimer(new TimerTask() {
                @Override
                public void run() {
                    disConnection(gatt);
                    errorCall(device.getAddress(), "Connection timeout", BleCode.CONNECTION_TIMEOUT);
                }
            }, 5000);

        }
    }


    /**
     * 断开连接
     */
    private void disConnection(BluetoothGatt gatt) {
        gatt.disconnect();
        startTimer(new TimerTask() {
            @Override
            public void run() {
                gatt.close();
            }
        }, 200);
    }


    //------------------蓝牙连接的回调--------------------------------//
    @Override
    public void onConnectionStateChange(BluetoothGatt gatt, int status, int newState) {
        super.onConnectionStateChange(gatt, status, newState);
        timer.cancel();
        if (status == BluetoothGatt.GATT_SUCCESS && newState == BluetoothProfile.STATE_CONNECTED) {
            gatt.discoverServices();
            startTimer(new TimerTask() {
                public void run() {
                    timer.cancel();
                    disConnection(gatt);
                    errorCall(gatt.getDevice().getAddress(), "Found service failure ", BleCode.GET_SERVICE_FAIL);
                }
            }, 10000);
        } else {
            gatt.disconnect();
            gatt.close();
            if (status == 8) {
                errorCall(gatt.getDevice().getAddress(), "Accidentally disconnected", BleCode.UNEXPECTED_DISCONNECT);
            } else if (status == BluetoothGatt.GATT_SUCCESS && newState == BluetoothProfile.STATE_DISCONNECTED) {
                if (!Ble.instance(context).getBlueAdapter().isEnabled())
                    errorCall(gatt.getDevice().getAddress(), "Bluetooth off", BleCode.BLUE_OFF);
            } else {
                errorCall(gatt.getDevice().getAddress(), "Connection device failed", BleCode.CONNECTION_FAIL);
            }
        }
    }

    @Override
    public void onServicesDiscovered(BluetoothGatt gatt, int status) {
        super.onServicesDiscovered(gatt, status);
        timer.cancel();
        processCall(gatt.getDevice().getAddress(), BleCode.CONNECT_SUCCESS);
        BluetoothGattService service = gatt.getService(UUID.fromString(Ble.SERVICES));
        BluetoothGattCharacteristic characteristic = service.getCharacteristic(UUID.fromString(Ble.BATTERY));
        processCall(gatt.getDevice().getAddress(), BleCode.SECURITY_CERTIFICATION_ONGOING);
        gatt.readCharacteristic(characteristic);
        startTimer(new TimerTask() {
            public void run() {
                timer.cancel();
                disConnection(gatt);
                errorCall(gatt.getDevice().getAddress(), "Read Data Timeout", BleCode.READ_OVERTIME);
            }
        }, 1500);

    }

    @Override
    public void onCharacteristicRead(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic, int status) {
        super.onCharacteristicRead(gatt, characteristic, status);
        timer.cancel();
        if (characteristic.getUuid().toString().equalsIgnoreCase(Ble.BATTERY)) {
            handlerFf01(gatt, characteristic, characteristic.getValue());
        } else if (characteristic.getUuid().toString().equalsIgnoreCase(Ble.KEY)) {
            deviceInfo.setKey(ByteToString.bytesToHexString(characteristic.getValue()));
            processCall(gatt.getDevice().getAddress(), BleCode.ACCESS_INFORMATION_COMPLETED);
            new Thread() {
                public void run() {
                    uploadDeviceInfo(gatt);
                }
            }.start();
        }

    }

    @Override
    public void onCharacteristicChanged(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
        super.onCharacteristicChanged(gatt, characteristic);
        timer.cancel();
        //接受到进入设置状态的通知
        if (characteristic.getUuid().toString().equalsIgnoreCase(Ble.BATTERY)) {
            deviceInfo = analyze(characteristic.getValue());
            if (deviceInfo.isSetup()) {
                processCall(gatt.getDevice().getAddress(), BleCode.DEVICE_SET_UP);
                readKey(gatt);
            }
        }


    }

    @Override
    public void onCharacteristicWrite(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic, int status) {
        super.onCharacteristicWrite(gatt, characteristic, status);
        timer.cancel();
        if (characteristic.getUuid().toString().equalsIgnoreCase(Ble.ADD_FINISH)) {
            processCall(gatt.getDevice().getAddress(), BleCode.ADD_SUCCESS);
            addSuccessCall(gatt.getDevice().getAddress(), BleCode.ADD_SUCCESS);
            disConnection(gatt);
        }

    }

    /**
     * 处理读取的ff01数据
     */
    private void handlerFf01(BluetoothGatt gatt, BluetoothGattCharacteristic characteristicFf01, byte[] data) {
        new Thread() {
            @Override
            public void run() {
                super.run();
                deviceInfo = analyze(data);
                int security = security(gatt.getDevice().getAddress(), deviceInfo);
                if (security == SECURITY_FAIL) {
                    disConnection(gatt);
                } else if (security == SECURITY_OK) {//验证成功立马读取数据
                    readKey(gatt);
                } else if (security == SECURITY_NO_SETTING_STATE) {//设置监听通知
                    startTimer(new TimerTask() {
                        public void run() {
                            timer.cancel();
                            disConnection(gatt);
                            errorCall(gatt.getDevice().getAddress(), "Wait to set state timeout", BleCode.WAIT_DEVICE_SET_UP_OVERTIME);
                        }
                    }, 1000);
                    monitorNotification(gatt, characteristicFf01);
                }

            }
        }.start();

    }


    /**
     * 设置接受ff01的通知
     */
    private void monitorNotification(BluetoothGatt gatt, BluetoothGattCharacteristic characteristicFf01) {
        boolean state = gatt.setCharacteristicNotification(characteristicFf01, true);
        if (state) {
            List<BluetoothGattDescriptor> descriptors = characteristicFf01.getDescriptors();
            if (descriptors == null || descriptors.size() == 0) {
                disConnection(gatt);
                errorCall(gatt.getDevice().getAddress(), "Serial authentication failed", BleCode.SERIAL_PORT_FAIL);
                return;
            }
            for (BluetoothGattDescriptor descriptor : descriptors) {
                descriptor.setValue(BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE);
                gatt.writeDescriptor(descriptor);
            }
        } else {
            disConnection(gatt);
            errorCall(gatt.getDevice().getAddress(), "Serial authentication failed", BleCode.SERIAL_PORT_FAIL);
        }
    }


    /**
     * 读取密钥
     */
    private void readKey(BluetoothGatt gatt) {
        processCall(gatt.getDevice().getAddress(), BleCode.ACCESS_INFORMATION);
        BluetoothGattService service = gatt.getService(UUID.fromString(Ble.SERVICES));
        BluetoothGattCharacteristic characteristicKey = service.getCharacteristic(UUID.fromString(Ble.KEY));
        gatt.readCharacteristic(characteristicKey);
        startTimer(new TimerTask() {
            public void run() {
                timer.cancel();
                disConnection(gatt);
                errorCall(gatt.getDevice().getAddress(), "Read Data Timeout", BleCode.READ_OVERTIME);
            }
        }, 1500);
    }


    /**
     * 解析发ff01的数据
     */
    private DeviceInfo analyze(byte[] ff01) {
        if (ff01 == null || ff01.length < 6)
            return null;
        DeviceInfo deviceInfo = new DeviceInfo();
        deviceInfo.setNewDevice(ff01[0] == 1);
        deviceInfo.setSetup(ff01[1] == 1);
        deviceInfo.setFirmwareVersion(ByteToString.bytesToHexString(new byte[]{ff01[2], ff01[3], ff01[4]}));
        deviceInfo.setBattery(ff01[6]);
        if (ff01.length <= 10) {
            return deviceInfo;
        }
        deviceInfo.setModel(ByteToString.bytesToHexString(new byte[]{ff01[7], ff01[8], ff01[9], ff01[10]}));
        byte[] other = new byte[ff01.length - 11];
        System.arraycopy(ff01, 11, other, 0, other.length);
        deviceInfo.setOther(ByteToString.bytesToHexString(other));
        return deviceInfo;
    }


    /**
     * 进行安全认证
     */
    private int security(String mac, DeviceInfo deviceInfo) {
        if (deviceInfo == null) {
            errorCall(mac, "Inconsistent equipment information", BleCode.FAILED_SECURITY_FAIL);
            return SECURITY_FAIL;
        }

        if (!deviceInfo.isNewDevice()) {
            errorCall(mac, "Device is not new.", BleCode.NO_NEW_DEVICE);
            return SECURITY_FAIL;
        }

        if (!NetWork.isNetworkConnected(context)) {
            errorCall(mac, "Current network not available.", BleCode.NO_NETWORK);
            return SECURITY_FAIL;
        }
        //网络认证设备
        Map<String, Object> map = new HashMap<>();
        map.put("deviceId", "Slock" + mac.replaceAll(":", ""));
        map.put("appId", appId);
        map.put("token", token);
        String clientData = HttpsClient.https("/device/get_verify_device", map);
        if (clientData == null) {
            errorCall(mac, "Failed to get service.", BleCode.SERVER_EXCEPTION);
            return SECURITY_FAIL;
        }
        try {
            JSONObject jsonObject = new JSONObject(clientData);
            if (jsonObject.getInt("err") != 0) {
                errorCall(mac, "Server authentication failed.", BleCode.SERVER_VERIFY_EXCEPTION);
                return SECURITY_FAIL;
            }
            checkCode = jsonObject.getJSONArray("data").getJSONObject(0).getInt("code");
        } catch (JSONException e) {
            errorCall(mac, "Failed to get service.", BleCode.SERVER_EXCEPTION);
            return SECURITY_FAIL;
        }
        if (!deviceInfo.isSetup()) {
            processCall(mac, BleCode.NO_DEVICE_SET_UP);
            return SECURITY_NO_SETTING_STATE;
        }
        processCall(mac, BleCode.SAFETY_CERTIFICATION_COMPLETED);
        return SECURITY_OK;
    }


    /**
     * 提交设备信息到服务器
     */
    private void uploadDeviceInfo(BluetoothGatt gatt) {
        processCall(gatt.getDevice().getAddress(), BleCode.ADDITIONS_BEING_COMPLETED);
        Map<String, Object> map = new HashMap<>();
        map.put("bleDeviceId", "Slock" + gatt.getDevice().getAddress().replaceAll(":", ""));
        map.put("checkCode", String.valueOf(checkCode));
        map.put("bleDeviceKey", deviceInfo.getKey());
        map.put("bleMac", gatt.getDevice().getAddress());
        map.put("bleDeviceModel", deviceInfo.getModel());
        map.put("bleVersion", deviceInfo.getFirmwareVersion());
        map.put("bleDeviceName", deviceInfo.getName() + DeviceManager.getInstance(context, appId, token).getDevicesNumber());//设备名字默认为
        map.put("bleDeviceBattery", String.valueOf(deviceInfo.getBattery()));//设备名字默认为
        //map["bleLongitude"]=val.battery.toString();
        //map["bleLatitude"]=val.battery.toString();
        String clientData = HttpsClient.https("/device/ble/add_ble_device", map);
        if (clientData == null) {
            disConnection(gatt);
            errorCall(gatt.getDevice().getAddress(), "Failed to get service.", BleCode.SERVER_EXCEPTION);
            return;
        }
        try {
            JSONObject jsonObject = new JSONObject(clientData);
            if (jsonObject.getInt("err") != 0) {
                disConnection(gatt);
                errorCall(gatt.getDevice().getAddress(), "Server authentication failed.", BleCode.SERVER_VERIFY_EXCEPTION);
                return;
            }
            processCall(gatt.getDevice().getAddress(), BleCode.ADD_FINISH);
            sendSuccessCommand(gatt);
        } catch (JSONException e) {
            e.printStackTrace();
            disConnection(gatt);
            errorCall(gatt.getDevice().getAddress(), "Failed to get service.", BleCode.SERVER_EXCEPTION);
            return;
        }


    }

    /**
     * 发送添加成功的指令
     */
    private void sendSuccessCommand(BluetoothGatt gatt) {
        BluetoothGattService service = gatt.getService(UUID.fromString(Ble.SERVICES));
        BluetoothGattCharacteristic characteristic = service.getCharacteristic(UUID.fromString(Ble.ADD_FINISH));
        characteristic.setValue(new byte[0x03]);
        gatt.writeCharacteristic(characteristic);
        startTimer(new TimerTask() {
            public void run() {
                timer.cancel();
                //如果出现这种情况代表服务器已经保存，设备没有反应，需要删除服务器设备重试
                disConnection(gatt);
                errorCall(gatt.getDevice().getAddress(), "Waiting for equipment confirmation timeout", BleCode.CONFIRMATION_TIMEOUT);
            }
        }, 1000);
    }


    /**
     * 创建一个统一的定时器整个过程中有且只能有一个定时器存在
     *
     * @param delay
     * @param timerTask
     */
    private void startTimer(TimerTask timerTask, int delay) {
        if (timer != null)
            timer.cancel();
        timer = new Timer();
        timer.schedule(timerTask, delay);
    }


    /**
     * 回调异常
     */
    private void errorCall(String deviceId, String err, int code) {
        if (addBleDeviceCall != null)
            addBleDeviceCall.error(deviceId != null ? "Slock" + deviceId.replaceAll(":", "") : null, err, code);
    }

    /**
     * 回调进度
     */
    private void processCall(String deviceId, int code) {
        if (addBleDeviceCall != null)
            addBleDeviceCall.addProcess(deviceId != null ? "Slock" + deviceId.replaceAll(":", "") : null, code);
    }


    /**
     * 回调扫描到的设备
     */
    private void scanDeviceCall(String deviceId, int rssi) {
        if (addBleDeviceCall != null)
            addBleDeviceCall.scanDevice(deviceId, rssi);
    }


    /**
     * 添加成功的回调
     */
    private void addSuccessCall(String deviceId, int code) {
        if (addBleDeviceCall != null)
            addBleDeviceCall.addSuccess("Slock" + deviceId.replaceAll(":", ""), code);
    }

}

File: ios/Classes/add/ble/AddBleDevice.swift
//
// Created by yyjjls on 2020/7/11.
//

import Foundation
import CoreBluetooth

//添加蓝牙设备
class AddBleDevice: NSObject, AddDevice, BleCall, CBPeripheralDelegate {
    private var appId: String?;
    private var token: String?;
    private var deviceId: String? = "";
    private var timer: Timer?;
    //安全认证通过
    private static let SECURITY_OK: Int = 0x00;
    //安全认证失败
    private static let SECURITY_FAIL: Int = 0x01;
    //沒有进入设置状态
    private static let SECURITY_NO_SETTING_STATE: Int = 0x02;

    private var addDeviceInfo: AddDeviceInfo?

    private var addBleDeviceCall: AddBleDeviceCall?
    //保存扫描到的设备
    private var scanDevicesMap = [String: CBPeripheral]();

    private static var addDevice: AddDevice?;

    //网络认证的权限码
    private var checkCode: String?;

    public static func getInstance(appId: String?, token: String?) -> AddDevice {
        if (addDevice == nil) {
            addDevice = AddBleDevice();
            (addDevice as! AddBleDevice).appId = appId;
            (addDevice as! AddBleDevice).token = token;
        }
        return addDevice!;
    }

    func addCall(addBleDeviceCall: AddBleDeviceCall) {
        self.addBleDeviceCall = addBleDeviceCall;
    }

    func scanDevice() {
        self.deviceId = "";
        scanDevicesMap.removeAll();
        Ble.getInstance.scan(serviceUUIDs: [Ble.SCAN2], bleCall: self);
        processCall(deviceId: deviceId, code: BleCode.SCANNING);
    }

    func stopDevice() {
        Ble.getInstance.stopScan();
    }

    func addDevice(deviceId: String?) {
        if (deviceId == nil) {
            return;
        }
        let deviceInfo = scanDevicesMap[deviceId!];
        if (deviceInfo == nil) {//如果没有换成代表用户是直接指定连接添加的设备
            self.deviceId = deviceId;
            scanDevicesMap.removeAll();
            Ble.getInstance.scan(serviceUUIDs: [Ble.SCAN2], bleCall: self);
            bleTimer(timeInterval: 5, aSelector: #selector(directConnectionOverTimer))
        } else {//已经扫描完成用户直接连接指定设备
            Ble.getInstance.connect(deviceInfo!);
        }
        processCall(deviceId: deviceId, code: BleCode.CONNECTING);
    }

    //当之前输入设备id进行添加，主要是出现扫描二维码添加，定时器
    @objc private func directConnectionOverTimer() {
        cancelAdd();
        errorCall(deviceId: deviceId!, err: "Device connection timeout", code: BleCode.CONNECTION_TIMEOUT);
    }

    func cancelAdd() {
        Ble.getInstance.stopScan();
        scanDevicesMap.forEach { key, value in
            if (value.state == CBPeripheralState.connected) {
                Ble.getInstance.disConnect(value);
            } else {
                Ble.getInstance.cancelConnection(value);
            }
        }
    }

    /*》》》》》》》》》》》蓝牙的回调《《《《《《《《《《《《《《《《《*/
    func belState(code: Int, msg: String) {
        if (code == BleCode.DEVICE_BLUE_OFF) {
            closeBleTimer(); //蓝牙关闭直接关闭定时器
        }
    }

    func scanDevice(central: CBCentralManager, peripheral: CBPeripheral, advertisementData: [String: Any], rssi: NSNumber) {
        if (peripheral.name != nil) {return;}
        if (deviceId! != "" && peripheral.name! == deviceId) { //连接指定的设备
            scanDevicesMap[peripheral.name!] = peripheral;
            Ble.getInstance.stopScan();
            Ble.getInstance.connect(peripheral);
        } else if (((peripheral.name?.contains("Slock")) != nil)) { //扫描附近的设备
            scanDevicesMap[peripheral.name!] = peripheral;
            processCall(deviceId: peripheral.name!, code: BleCode.SCAN_ADD_DEVICE_INFO);
            scanDeviceCall(deviceId: peripheral.name!, rssi: Int(truncating: rssi));
            Ble.getInstance.stopScan();
            bleTimer(timeInterval: 5, aSelector: #selector(connectionOverTimer))
        }
    }

    //连接设备超时
    @objc private func connectionOverTimer() {
        cancelAdd();
        errorCall(deviceId: deviceId!, err: "Device connection timeout", code: BleCode.CONNECTION_TIMEOUT);
    }

    func error(code: Int, error: String) {
        closeBleTimer();
        errorCall(deviceId: deviceId!, err: error, code: code);
    }

    func connect(central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        closeBleTimer();
        processCall(deviceId: peripheral.name!, code: BleCode.CONNECT_SUCCESS);
        peripheral.delegate = self;
        peripheral.discoverServices([Ble.SERVICES])
        processCall(deviceId: peripheral.name!, code: BleCode.SECURITY_CERTIFICATION_ONGOING);
    }


    func disconnect(central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        if (error != nil) {
            errorCall(deviceId: peripheral.name!, err: "Bluetooth accidental disconnect", code: BleCode.UNEXPECTED_DISCONNECT)
        }
    }

    //发现服务
    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        if (error != nil) {
            errorCall(deviceId: peripheral.name!, err: "Discover service failure", code: BleCode.GET_SERVICE_FAIL)
            Ble.getInstance.disConnect(peripheral);
            return;
        }
        peripheral.discoverCharacteristics(nil, for: peripheral.services![0])
    }

    //发现特征值
    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        if (error != nil) {
            errorCall(deviceId: peripheral.name!, err: "Discover Characteristics failure", code: BleCode.GET_CHARACTERISTIC_FAIL)
            Ble.getInstance.disConnect(peripheral);
            return;
        }
        //设备电量信息等
        peripheral.readValue(for: Ble.getInstance.getCharacteristic(services: peripheral.services![0], uuid: Ble.BATTERY)!);

    }

    //读取到的值
    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        if (error != nil) {
            errorCall(deviceId: peripheral.name!, err: "Failed to read data", code: BleCode.READ_DATA_FAIL)
            Ble.getInstance.disConnect(peripheral);
            return;
        }
        if (characteristic.uuid.isEqual(Ble.BATTERY)) {
            handlerFf01(peripheral: peripheral, didUpdateValueFor: characteristic);
        } else if (characteristic.uuid.isEqual(Ble.KEY)) {
            addDeviceInfo?.key = characteristic.value!.toHexString();
            processCall(deviceId: peripheral.name!, code: BleCode.ACCESS_INFORMATION_COMPLETED);
            uploadDeviceInfo(peripheral: peripheral);
        }
    }

    //通知设置成功的回调
    func peripheral(_ peripheral: CBPeripheral, didUpdateNotificationStateFor characteristic: CBCharacteristic, error: Error?) {
        if (error != nil) {
            errorCall(deviceId: peripheral.name!, err: "Failed to listen for notification", code: BleCode.NOTIFICATION_DATA_FAIL);
            Ble.getInstance.disConnect(peripheral);
            return;
        }
        // processCall(deviceId: peripheral.name!, code: BleCode.ACCESS_INFORMATION_COMPLETED);
    }

    //写入值成功
    func peripheral(_ peripheral: CBPeripheral, didWriteValueFor characteristic: CBCharacteristic, error: Error?) {
        if (error != nil) {
            errorCall(deviceId: peripheral.name!, err: "Failed to write data", code: BleCode.WRITE_DATA_FAIL)
            Ble.getInstance.disConnect(peripheral);
            return;
        }
        if (characteristic.uuid.isEqual(Ble.ADD_FINISH)) {
            Ble.getInstance.disConnect(peripheral);
            //添加成功
            processCall(deviceId: peripheral.name!, code: BleCode.ADD_SUCCESS);
            addSuccessCall(deviceId: peripheral.name!, code: BleCode.ADD_SUCCESS);
        }
    }

    /**
     * 处理读取的ff01数据
     */
    private func handlerFf01(peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic) {
        addDeviceInfo = analyze(ff01: characteristic.value!);
        let state = security(deviceId: peripheral.name!, addDeviceInfo: addDeviceInfo);
        if (state == AddBleDevice.SECURITY_FAIL) {//安全认证失败直接结束
            Ble.getInstance.disConnect(peripheral);
            return;
        } else if (state == AddBleDevice.SECURITY_NO_SETTING_STATE) { //等待进入设置状态
            peripheral.setNotifyValue(true, for: Ble.getInstance.getCharacteristic(services: peripheral.services![0], uuid: Ble.BATTERY)!);
        } else if (state == AddBleDevice.SECURITY_OK) {//安全认证通过去读取key
            processCall(deviceId: peripheral.name!, code: BleCode.ACCESS_INFORMATION);
            peripheral.readValue(for: Ble.getInstance.getCharacteristic(services: peripheral.services![0], uuid: Ble.KEY)!);

        }
    }

    /**
    * 解析发ff01的数据
    */
    private func analyze(ff01: Data) -> AddDeviceInfo? {
        if (ff01.count < 6) {
            return nil;
        }
        var addDeviceInfo = AddDeviceInfo();
        addDeviceInfo.isNewDevice = (ff01[0] == 1);
        addDeviceInfo.isSetup = (ff01[1] == 1);
        addDeviceInfo.firmwareVersion = ff01.subdata(in: 2..<5).toHexString();
        addDeviceInfo.battery = Int(ff01[6]);
        if (ff01.count <= 10) {
            return addDeviceInfo;
        }
        addDeviceInfo.model = ff01.subdata(in: 7..<11).toHexString();
        if (ff01.count >= 11) {
            addDeviceInfo.other = ff01.subdata(in: 11..<ff01.count).toHexString();
        }
        return addDeviceInfo;
    }

    /**
    * 进行安全认证
    */
    private func security(deviceId: String?, addDeviceInfo: AddDeviceInfo?) -> Int {
        if (addDeviceInfo == nil) {
            errorCall(deviceId: deviceId!, err: "Inconsistent equipment information", code: BleCode.FAILED_SECURITY_FAIL);
            return AddBleDevice.SECURITY_FAIL;
        }

        if (!addDeviceInfo!.isNewDevice) {
            errorCall(deviceId: deviceId!, err: "Device is not new.", code: BleCode.NO_NEW_DEVICE);
            return AddBleDevice.SECURITY_FAIL;
        }
        if (!addDeviceInfo!.isSetup) {
            processCall(deviceId: deviceId, code: BleCode.NO_DEVICE_SET_UP);
            return AddBleDevice.SECURITY_NO_SETTING_STATE;
        }

//        if (!NetWork.isNetworkConnected(context)) {
//            errorCall(mac, "Current network not available.", BleCode.NO_NETWORK);
//            return SECURITY_FAIL;
//        }
        //网络认证设备
        let paramDic: [String: String] = ["deviceId": deviceId!, "appId": appId!, "token": token!]
        let clientData: NSDictionary? = HttpsClient.POSTAction(urlStr: "/device/get_verify_device", param: paramDic);
        if (clientData == nil) {
            errorCall(deviceId: deviceId!, err: "Failed to get service.", code: BleCode.SERVER_EXCEPTION);
            return AddBleDevice.SECURITY_FAIL;
        }
        if (clientData!["err"] as! Int != 0) {
            errorCall(deviceId: deviceId!, err: "Server authentication failed.", code: BleCode.SERVER_VERIFY_EXCEPTION);
            return AddBleDevice.SECURITY_FAIL;
        }
        checkCode = (((((clientData!["data"] as! NSArray)[0]) as! NSDictionary)["code"]) as! String);
        processCall(deviceId: deviceId, code: BleCode.SAFETY_CERTIFICATION_COMPLETED);
        return AddBleDevice.SECURITY_OK;
    }

    /**
    * 提交设备信息到服务器
    */
    private func uploadDeviceInfo(peripheral: CBPeripheral) {
        processCall(deviceId: peripheral.name, code: BleCode.ADDITIONS_BEING_COMPLETED);
//        if (AppLocation.getLocation() != null) {
//        map.put("bleLongitude", AppLocation.getLocation().getLongitude());
//        map.put("bleLatitude", AppLocation.getLocation().getLatitude());
//        map.put("blePosition", AppLocation.getLocationAddress(context, AppLocation.getLocation()));
//        }
        let deviceName = peripheral.name!
        var hexMac = deviceName.suffix(12);
        let per = 2;
        let spaceCount = Int(hexMac.count / per)
        for i in 1..<spaceCount {
            hexMac.insert(contentsOf: ":", at: hexMac.index(hexMac.startIndex, offsetBy: per * i + i - 1))
        }
        let deviceNumber = DeviceManager.getInstance(appId: appId!, token: token!).getDevicesNumber();
        let paramDic: [String: String] = ["bleDeviceId": deviceName, "appId": appId!,
                                          "token": token!, "checkCode": checkCode!, "bleDeviceKey": addDeviceInfo!.key,
                                          "bleMac": String(hexMac), "bleDeviceModel": addDeviceInfo!.model, "bleVersion": addDeviceInfo!.firmwareVersion,
                                          "bleDeviceName": addDeviceInfo!.name + String(deviceNumber)//设备名字默认为
                                          , "bleDeviceBattery": String(addDeviceInfo!.battery)]

        let clientData: NSDictionary? = HttpsClient.POSTAction(urlStr: "/device/ble/add_ble_device", param: paramDic);
        if (clientData == nil) {
            Ble.getInstance.disConnect(peripheral);
            errorCall(deviceId: peripheral.name!, err: "Failed to get service.", code: BleCode.SERVER_EXCEPTION);
            return;
        }

        if (clientData!["err"] as! Int != 0) {
            Ble.getInstance.disConnect(peripheral);
            errorCall(deviceId: peripheral.name!, err: "Server authentication failed.", code: BleCode.SERVER_VERIFY_EXCEPTION);
            return;
        }
        sendSuccessCommand(peripheral: peripheral);
    }

    /**
        * 发送添加成功的指令
        */
    private func sendSuccessCommand(peripheral: CBPeripheral) {
        peripheral.writeValue(Data.init(hex: "0x03"), for: Ble.getInstance.getCharacteristic(services: peripheral.services![0], uuid: Ble.ADD_FINISH)!, type: CBCharacteristicWriteType.withResponse);
        processCall(deviceId: peripheral.name!, code: BleCode.ADD_FINISH);
        //errorCall(gatt.getDevice().getAddress(), "Waiting for equipment confirmation timeout", BleCode.CONFIRMATION_TIMEOUT);
    }


/**
* 回调异常
*/
    private func errorCall(deviceId: String, err: String, code: Int) {
        if (addBleDeviceCall != nil) {
            addBleDeviceCall?.error(deviceId: deviceId, err: err, code: code);
        }
    }

    /**
    * 回调进度
    */
    private func processCall(deviceId: String?, code: Int) {
        if (addBleDeviceCall != nil) {
            addBleDeviceCall?.addProcess(deviceId: deviceId, code: code);
        }
    }


    /**
    * 回调扫描到的设备
    */
    private func scanDeviceCall(deviceId: String, rssi: Int) {
        if (addBleDeviceCall != nil) {
            addBleDeviceCall?.scanDevice(deviceId: deviceId, rssi: rssi);
        }
    }


    /**
    * 添加成功的回调
    */
    private func addSuccessCall(deviceId: String, code: Int) {
        if (addBleDeviceCall != nil) {
            addBleDeviceCall?.addSuccess(deviceId: deviceId, code: code);
        }
    }

    //定时器
    private func bleTimer(timeInterval: Double, aSelector: Selector) {
        timer = Timer.scheduledTimer(timeInterval: timeInterval, target: self, selector: aSelector, userInfo: nil, repeats: false);
    }

    //关闭定时
    private func closeBleTimer() {
        if (timer != nil) {
            timer?.invalidate()
        }
    }

}
